crates/libs/nuts/src/nut12/mod.rs

```diff
@@ -0,0 +1,347 @@
+
+    // 2. Add the result to `s_sk` (which is `r`) in-place. `s_sk` now holds `r + e*a`.
+    let s_sk = s_sk.add_tweak(&ea_scalar)?;
+
+    let s_scalar = Scalar::from_be_bytes(s_sk.secret_bytes()).expect("Valid scalar");
+
+    Ok((
+        hex::encode(e_scalar.to_be_bytes()),
+        hex::encode(s_scalar.to_be_bytes()),
+    ))
+}
+
+/// Verify DLEQ proof during minting/swapping
+/// Alice verifies the mint's signature on her blinded message
+pub fn verify_dleq_proof_alice(
+    e_hex: &str,
+    s_hex: &str,
+    mint_public_key: &PublicKey,
+    blinded_message: &PublicKey,
+    blind_signature: &PublicKey,
+) -> Result<bool, Error> {
+    let e_scalar = Scalar::from_be_bytes(
+        hex::decode(e_hex)?
+            .try_into()
+            .map_err(|_| Error::InvalidDleqProof)?,
+    )?;
+    let s_scalar = Scalar::from_be_bytes(
+        hex::decode(s_hex)?
+            .try_into()
+            .map_err(|_| Error::InvalidDleqProof)?,
+    )?;
+
+    // R1 = sG - eA
+    let s_sk = SecpSecretKey::from_slice(&s_scalar.to_be_bytes())?;
+    let s_g = SecpPublicKey::from_secret_key(&SECP256K1, &s_sk);
+    let e_a = mint_public_key.mul_tweak(&SECP256K1, &e_scalar)?;
+    let e_a_neg = e_a.negate(&SECP256K1);
+    let r1_computed = PublicKey::from(s_g.combine(&e_a_neg)?);
+
+    // R2 = sB' - eC'
+    let s_b_prime = blinded_message.mul_tweak(&SECP256K1, &s_scalar)?;
+    let e_c_prime = blind_signature.mul_tweak(&SECP256K1, &e_scalar)?;
+    let e_c_prime_neg = e_c_prime.negate(&SECP256K1);
+    let r2_computed = PublicKey::from(s_b_prime.combine(&e_c_prime_neg)?);
+
+    let e_computed_bytes =
+        dhke::hash_e([r1_computed, r2_computed, *mint_public_key, *blind_signature]);
+
+    // Return Ok with the boolean result of the comparison.
+    Ok(e_scalar.to_be_bytes() == e_computed_bytes)
+}
+
+/// Verify DLEQ proof for a received token (Carol's case).
+pub fn verify_dleq_proof_carol(
+    e_hex: &str,
+    s_hex: &str,
+    r_hex: &str,
+    mint_public_key: &PublicKey, // A
+    secret: &[u8],
+    unblinded_signature: &PublicKey, // C
+) -> Result<bool, Error> {
+    let r_scalar = Scalar::from_be_bytes(
+        hex::decode(r_hex)?
+            .try_into()
+            .map_err(|_| Error::InvalidDleqProof)?,
+    )?;
+
+    // Reconstruct B' = Y + rG
+    let y = dhke::hash_to_curve(secret)?;
+    let r_sk = SecpSecretKey::from_slice(&r_scalar.to_be_bytes())?;
+    let r_g = SecpPublicKey::from_secret_key(&SECP256K1, &r_sk);
+    let blinded_message = PublicKey::from(y.combine(&r_g)?);
+
+    // Reconstruct C' = C + rA
+    let r_a = mint_public_key.mul_tweak(&SECP256K1, &r_scalar)?;
+    let blinded_signature = PublicKey::from(unblinded_signature.combine(&r_a)?);
+
+    verify_dleq_proof_alice(
+        e_hex,
+        s_hex,
+        mint_public_key,
+        &blinded_message,
+        &blinded_signature,
+    )
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Add tests where things go wrong. Invalid proofs, etc.

Move the test in a `nuts/src/nut12/tests.rs` file.
There is many of them and it takes a lot of space